src/devices/machine/i82586.h

```diff
@@ -9,28 +9,254 @@
+	virtual void cu_execute();
+	virtual bool cu_iasetup() = 0;
+	virtual bool cu_configure() = 0;
+	virtual bool cu_mcsetup() = 0;
+	virtual bool cu_transmit(u32 command) = 0;
+	virtual bool cu_tdreflect() = 0;
+	virtual bool cu_dump() = 0;
+
+	// receive unit
+	virtual bool accept_filter(u8 *mac);
+	virtual void ru_execute(u8 *buf, int length) = 0;
+
+	// helpers
+	void set_irq(bool irq);
+	virtual u32 address(u32 base, int offset, int address) = 0;
+	virtual u32 address(u32 base, int offset, int address, u16 empty) = 0;
+	static u32 compute_crc(u8 *buf, int length, bool crc32);
+	int fetch_bytes(u8 *buf, u32 src, int length);
+	int store_bytes(u32 dst, u8 *buf, int length);
+	void dump_bytes(u8 *buf, int length);
+
+	// device_* members
 	const address_space_config m_space_config;
 	address_space *m_space;
 
 	devcb_write_line m_out_irq;
+	static const device_timer_id CU_TIMER = 0;
+	static const device_timer_id RU_TIMER = 1;
+	emu_timer *m_cu_timer;
+	emu_timer *m_ru_timer;
+
+	// interrupt state
+	bool m_cx;          // command executed (with interrupt)
+	bool m_fr;          // frame received
+	bool m_cna;         // command unit became inactive
+	bool m_rnr;         // receive unit became not ready
+	bool m_irq_state;
+	bool m_initialised;
+
+	// receive/command unit state
+	cu_state m_cu_state; // command unit state
+	ru_state m_ru_state; // receive unit state
+	u32 m_scp_address;   // system configuration pointer
+	u32 m_scb_base;      // system control block base (segmented modes)
+	u32 m_scb_address;   // system control block address (linear mode)
+	u32 m_scb_cs;        // current scb command and status
+	u32 m_cba;           // current command block address
+	u32 m_rfd;           // current receive frame descriptor address
+
+	// addresses
+	u64 m_mac;                    // current individual address
+	std::vector<u64> m_mac_multi; // multicast address filter list
+
+	u8 m_lb_buf[1528]; // storage for loopback frames
+	int m_lb_length;   // length of loopback frame
+
+	// configure parameters
+	enum loopback_mode
+	{
+		LOOPBACK_NONE     = 0x00,
+		LOOPBACK_INTERNAL = 0x40, // not sure which is which
+		LOOPBACK_EXTERNAL = 0x80,
+		LOOPBACK_BOTH     = 0xc0  // internal loopback for 82586, external loopback for 82596
+	};
+	virtual u8 cfg_get(int offset) = 0;
+	virtual void cfg_set(int offset, u8 data) = 0;
+
+	bool cfg_save_bad_frames()   { return BIT(cfg_get(2), 7); }
+	u8 cfg_address_length()      { return (cfg_get(3) & 0x7) % 7; } // "NOTE: 7 is interpreted as 0"
+	bool cfg_no_src_add_ins()    { return BIT(cfg_get(3), 3); }
+	loopback_mode cfg_loopback_mode()        { return (loopback_mode)(cfg_get(3) & LOOPBACK_BOTH); }
+	bool cfg_promiscuous_mode()  { return BIT(cfg_get(8), 0); }
+	bool cfg_broadcast_disable() { return BIT(cfg_get(8), 1); }
+	bool cfg_no_crc_insertion()  { return BIT(cfg_get(8), 4); }
+	bool cfg_crc16()             { return BIT(cfg_get(8), 5); }
+	u8 cfg_min_frame_length()    { return cfg_get(10); }
+
+private:
```

Comment:
Only to make it a bit easier to see the different byte/bit accesses line up, not required.